crates/matrix-sdk-crypto/src/olm/account.rs

```diff
@@ -1128,32 +1122,77 @@ impl Account {
+            let sessions: &mut Vec<Session> = &mut *sessions.lock().await;
+            // Try to decrypt the message using each Session we share with the
+            // given curve25519 sender key.
+            for session in sessions {
+                if let Ok(p) = session.decrypt(message).await {
+                    // success!
+                    return Ok((SessionType::Existing(session.clone()), p));
+                } else {
+                    // An error here is completely normal, after all we don't know
+                    // which session was used to encrypt a message. We will log a
+                    // warning if no session was able to decrypt the message.
+                    continue;
+                }
+            }
+        }
+
+        let session_ids = if let Some(sessions) = existing_sessions {
+            sessions.lock().await.iter().map(|s| s.session_id().to_owned()).collect()
+        } else {
+            vec![]
+        };
+
+        warn!(?session_ids, "Failed to decrypt a non-pre-key message with all available sessions");
+        return Err(OlmError::SessionWedged(sender.to_owned(), sender_key));
+    }
 
-        let Some(sessions) = s else {
-            // We don't have any existing sessions, return early.
-            return Ok(None);
+    /// Try to decrypt a prekey Olm message.
+    ///
+    /// If we already know about the session corresponding to the message, use
+    /// it to decrypt the message. Otherwise, create a new Olm session.
+    async fn decrypt_prekey_olm_message(
+        &self,
+        store: &Store,
+        sender: &UserId,
+        sender_key: Curve25519PublicKey,
+        message: &OlmMessage,
+    ) -> OlmResult<(SessionType, String)> {
+        let OlmMessage::PreKey(ref prekey_message) = message else {
+            panic!("decrypt_prekey_olm_message called with a non-prekey message");
         };
 
-        let mut decrypted: Option<(Session, String)> = None;
+        // First try to decrypt using an existing session.
+        let existing_sessions = store.get_sessions(&sender_key.to_base64()).await?;
+        if let Some(ref sessions) = existing_sessions.as_ref() {
+            // Try to find the right session
+            for session in &mut *sessions.lock().await {
+                if prekey_message.session_id() != session.session_id() {
+                    // wrong session
+                    continue;
+                }
 
-        // Try to decrypt the message using each Session we share with the
-        // given curve25519 sender key.
-        for session in &mut *sessions.lock().await {
-            if let Ok(p) = session.decrypt(message).await {
-                decrypted = Some((session.clone(), p));
-                break;
-            } else if let OlmMessage::PreKey(message) = message {
-                if message.session_id() == session.session_id() {
+                return if let Ok(p) = session.decrypt(message).await {
```

Comment:
Maybe just me, but I find `return if let` / `return match` a bit weird to read / easy to miss, I'd personally prefer either two returns, on in each branch, or a `let` binding for the result of the `if let / else` expression that's then returned.

Reply:
I agree. I started out with separate `return`s, but my IDE wanted me to hoist them outside the `if`. I'll put them back.